### PR TITLE
VACMS 3223:  User Role dimension for the CMS Google Analytics data

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

See #3223.

I do not currently have access to Google Tag Manager, so I can't add the necessary mapping.  

I'm setting `currentUserRoles` to an array of user role machine names, e.g.

```
    | role                | roles                                   |
    | "anonymous"         | '["anonymous"]'                         |
    | "authenticated"     | '["authenticated"]'                     | 
    | "content_admin"     | '["authenticated","content_admin"]'     | 
    | "content_publisher" | '["authenticated","content_publisher"]' |
    | "administrator"     | '["authenticated","administrator"]'     |
```

I also added a Behat test and some code to `\CustomDrupal\FeatureContext` to support the test and any other tests we want to perform on the JS Drupal Settings or on the GTM dataLayer.  Please let me know if this is inappropriate or should've been done in a different place.

## Testing done
I added a Behat test and some supporting code and ran it for the roles listed above.

## Screenshots
Screen Shot 2020-10-26 at 12 51 51 PM

## QA steps
As any user
1. Log in and view any page in Chrome or some other browser with a developer console.
2. Open developer console and type `dataLayer[1].currentUserRoles`.
3. Verify that the console responds with a comma-separated list of user roles.

## Definition of Done
- [x] Mapping needs to be added in GTM to set the appropriate GA custom dimension.  I have access to GA but not GTM yet.

### CODE REVIEW

**Comment by ndouglas** [comment]
@ethanteague This change is two lines of code or so, but I thought you might have some thoughts about the way I did the tests, whether they're worth doing that way, whether I should expand them/change them/rip them out completely, etc.  Probably (?) can't be merged until the GTM/GA work is done elsewhere, but thought it might be worth reviewing anyway.

File: tests/behat/features/google_tag_manager.feature [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+Feature: Google Tag Manager dataLayer values are correct
+  In order to confirm that Google Tag Manager data is recorded properly
+  As anyone involved in the project
+  I need to see the correct dataLayer values
+
+  @user
+  Scenario Outline: Users should see an appropriate list of roles in the dataLayer.
+    Given I am logged in as a user with the <role> role
+    And I am on "/user"
+    Then google tag manager id is "GTM-WQ3DLLB"
```
Comment (ethanteague) [inline tests/behat/features/google_tag_manager.feature:11]: @ndouglas I don't think this step / function is necessary - googleTagManagerIdIs step logic should be enough.

Comment (ndouglas) [reply tests/behat/features/google_tag_manager.feature:11]: Fair enough 👍

**Review by indytechcook — DISMISSED** [review]

File: docroot/modules/custom/va_gov_backend/js/gtm_tag_trackers.js [hunk 2]
```diff
@@ -55,6 +55,7 @@
         'contentTitle': settings.gtm_data.contentTitle ? settings.gtm_data.contentTitle : null,
         'contentType': settings.gtm_data.contentType ? settings.gtm_data.contentType : null,
         'contentOwner': settings.gtm_data.contentOwner ? settings.gtm_data.contentOwner : null,
+        'currentUserRoles': settings.gtm_data.currentUserRoles ? settings.gtm_data.currentUserRoles : null,
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/js/gtm_tag_trackers.js:58]:
```suggestion
        'userRoles': settings.gtm_data.userRoles ? settings.gtm_data.userRoles : null,
```

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 3]
```diff
@@ -1023,6 +1023,8 @@ function _va_gov_backend_gtm_settings() {
     }
 
   }
+  $user_roles = \Drupal::currentUser()->getRoles() ?? ['anonymous'];
+  $sections['currentUserRoles'] = json_encode($user_roles);
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:1027]:
```suggestion
  $sections['userRoles'] = json_encode($user_roles);
```

File: tests/behat/features/google_tag_manager.feature [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    | "authenticated"     | '["authenticated"]'                     | 
+    | "content_admin"     | '["authenticated","content_admin"]'     | 
+    | "content_publisher" | '["authenticated","content_publisher"]' |
+    | "administrator"     | '["authenticated","administrator"]'     |
+
+  @user
+  Scenario: Google Tag Manager should indicate anonymous users.
+    Given I am an anonymous user
+    And I am on "/node/2"
+    Then google tag manager data layer value for "currentUserRoles" should be '["anonymous"]'
```
Comment (oksana-c) [inline tests/behat/features/google_tag_manager.feature:23]:
```suggestion
    Then google tag manager data layer value for "userRoles" should be '["anonymous"]'
```

File: docroot/modules/custom/va_gov_backend/js/gtm_tag_trackers.js (hunk 2 above)
Comment (oksana-c) [reply docroot/modules/custom/va_gov_backend/js/gtm_tag_trackers.js:58]: changing `currentUserRoles` to `userRoles` to match GA/GTM naming. Having `current` in the var name is not necessary for GA context.

**Review by oksana-c — CHANGES_REQUESTED** [review]

**Review by oksana-c — APPROVED** [review]
code/tests look great.
disregard my request to filter out `authenticated` role. After all,  it has its use. When new users are registered in the CMS they go without having a role for some time, so we could track their info separately.

**Comment by oksana-c** [comment]
prepped GA and GTM, and will coordinate updated container release once this baby goes out to prod ⚡

**Pull request merged**